src/peft/tuners/lora/layer.py

```diff
@@ -2004,10 +2008,34 @@ def _activate_lora(self, active_adapters: list[str]):
         try:
             yield
         finally:
-            nn.utils.parametrize.remove_parametrizations(
-                self.base_layer, self.parameter_name, leave_parametrized=False
+            self._remove_parametrizations()
+
+    def _remove_parametrizations(self):
+        # Remove the parametrization of this specific parameter
+        base_layer = self.get_base_layer()
+        parameter_name = self.parameter_name
+        if parameter_name not in base_layer.parametrizations:
+            raise ValueError(
+                "Something went wrong, please report this issue on PEFT: https://github.com/huggingface/peft/issues"
             )
 
+        if len(base_layer.parametrizations) == 1:
+            # last parametrization, we can safely remove it completely
+            nn.utils.parametrize.remove_parametrizations(base_layer, parameter_name, leave_parametrized=False)
+        else:
+            # More than one parametrization, only remove this specific one. Unfortunately, torch does not implement a
+            # way to only remove a single parametrization, so we implement this ourselves and follow original torch code
+            # closely:
+            # see https://github.com/pytorch/pytorch/blob/7d296d5c19750cecd82e2b95f6fb0f8dd918282e/torch/nn/utils/parametrize.py#L731-L737
+
+            original = base_layer.parametrizations[parameter_name].original
+            # Delete the property that manages the parametrization
+            delattr(base_layer.__class__, parameter_name)
+            # Delete the ParametrizationList
+            del base_layer.parametrizations[parameter_name]
+            # Restore the parameter / buffer into the main class
+            _register_parameter_or_buffer(base_layer, parameter_name, original)
+
```

Comment:
Maybe I'm misunderstanding the comment but AFAICS [`remove_parametrizations`](https://github.com/pytorch/pytorch/blob/v2.7.0/torch/nn/utils/parametrize.py#L652) only removes the parametrization for `tensor_name`, in this case `self.parameter_name`? Has this changed between pytorch versions?

Reply:
@githubnemo That sounds right. To expand on this:`len(base_layer.parametrizations) == 1` tells us that only one parameter has a parametrization, and based on the check prior to it, we know that it is `parameter_name`. However it doesn't tell us how many parametrizations are applied to that parameter; i.e. we just know we have only one `ParametrizationList` which may have length >= 1.

Since we're interested in just the parametrizations on `self.parameter_name` we should instead check `len(base_layer.parametrizations[parameter_name]) == 1`. If it is, assume the `remove_parametrizations` API can be safely used, even if we've got more than one target parameter on the module.

Otherwise, we shouldn't be deleting the entire ParametrizationList; just the specific `_LoraParameterProxy`.

Reply:
Yes, you are both right, this doesn't really make sense as is. For now, I "reverted" to `nn.utils.parametrize.remove_parametrizations`, which is enough for this PR, and added a `# TODO` for your case, Matthew. Let's fix that in a separate PR.